### PR TITLE
Remove `node.mode` and `node.local` settings

### PR DESCRIPTION
**Pull request by s1monw**
Today `node.mode` and `node.local` serve almost the same purpose, they
are a shortcut for `discovery.type` and `transport.type`. If `node.local: true`
or `node.mode: local` is set elasticsearch will start in _local_ mode which means
only nodes within the same JVM are discovered and a non-network based transport
is used. The _local_ mode it only really used in tests or if nodes are embedded.
For both, embedding and tests explicit configuration via `discovery.type` and `transport.type`
should be preferred.

This change removes all the usage of these settings and by-default doesn't
configure a default transport implementation since netty is now a module. Yet, to make
the user experience flawless, plugins or modules can set a `http.type.default` and
`transport.type.default`. Plugins set this via `PluginService#additionalSettings()`
which enforces _set-once_ which prevents node startup if set multiple times. This means
that our distributions will just startup with netty transport since it's packaged as a
module unless `transport.type` or `http.transport.type` is explicitly set.

This change also found a bunch of bugs since several `NamedWriteables` were not registered if a
transport client is used. Now that we don't rely on the `node.mode` leniency which is inherited
instead of using explicit settings, `TransportClient` uses `AssertingLocalTransport` which detects
these problems since it serializes all messages.

### CODE REVIEW

**Comment by s1monw** [comment]
@nik9000 FYI I had to fix some NamedWriteable regsitration here since they are relevant in the transport client case... now tests are failing too, we should stiff consider some unitttests for this stuff

File: modules/transport-netty/src/main/java/org/elasticsearch/transport/NettyPlugin.java [hunk 1]
```diff
@@ -85,6 +85,13 @@ public NettyPlugin(Settings settings) {
         );
     }
 
+    @Override
+    public Settings additionalSettings() {
+        return Settings.builder().put(super.additionalSettings())
```
Comment (rjernst) [inline modules/transport-netty/src/main/java/org/elasticsearch/transport/NettyPlugin.java]: no need for the super call, it is just the abstract Plugin that returns an empty list

**Comment by rjernst** [comment]
LGTM. Maybe as a followup we should move local transport/discovery to the test framework?

**Comment by s1monw** [comment]
> LGTM. Maybe as a followup we should move local transport/discovery to the test framework?

I agree I will open a followup

**Comment by javanna** [comment]
Relates to #16234, actually it should close it as far as I understand.

**Comment by s1monw** [comment]
@javanna ack I will close it in the final commit

**Pull request merged**